src/let_if_seq.rs

```diff
@@ -0,0 +1,141 @@
+/// let foo = if bar() {
+///     42;
+/// } else {
+///     0;
+/// };
+///
+/// let baz = if bar() {
+///     Some(42);
+/// } else {
+///     None
+/// };
+/// ```
+declare_lint! {
+    pub USELESS_LET_IF_SEQ,
+    Warn,
+    "TODO"
+}
+
+#[derive(Copy,Clone)]
+pub struct LetIfMut;
+
+impl LintPass for LetIfMut {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(USELESS_LET_IF_SEQ)
+    }
+}
+
+impl LateLintPass for LetIfMut {
+    fn check_block(&mut self, cx: &LateContext, block: &Block) {
+        let mut it = block.stmts.iter().peekable();
+        while let Some(ref stmt) = it.next() {
+            if_let_chain! {[
+                let StmtDecl(ref decl, _) = stmt.node,
+                let DeclLocal(ref decl) = decl.node,
+                let PatKind::Ident(mode, ref name, None) = decl.pat.node,
+                let Some(expr) = it.peek(),
+                let StmtExpr(ref if_, _) = expr.node,
+                let ExprIf(ref cond, ref then, ref else_) = if_.node,
+                let Some(value) = check_assign(name.node, then),
+            ], {
+                let span = codemap::mk_sp(stmt.span.lo, if_.span.hi);
+
+                let default = if let Some(ref else_) = *else_ {
+                    if let ExprBlock(ref else_) = else_.node {
+                        if let Some(default) = check_assign(name.node, else_) {
+                            default
+                        } else if let Some(ref default) = decl.init {
+                            &*default
+                        } else {
+                            return;
+                        }
+                    } else {
+                        return;
+                    }
+                } else if let Some(ref default) = decl.init {
+                    &*default
+                } else {
+                    return;
+                };
+
+                let mutability = match mode {
+                    BindByRef(MutMutable) | BindByValue(MutMutable) => "<mut> ",
+                    _ => "",
+                };
+
+                // FIXME: this should not suggest `mut` if we can detect that the variable is not
+                // use mutably after the `if`
+
+                let sug = format!(
+                    "let {mut}{name} = if {cond} {{{then} {value} }} else {{{else} {default} }};",
+                    mut=mutability,
+                    name=name.node.name,
+                    cond=snippet(cx, cond.span, "_"),
+                    then=if then.stmts.len() > 1 { " ..;" } else { "" },
+                    else=if else_.is_some() { " ..;" } else { "" },
+                    value=snippet(cx, value.span, "<value>"),
+                    default=snippet(cx, default.span, "<default>"),
+                );
+                span_lint_and_then(cx,
+                                   USELESS_LET_IF_SEQ,
+                                   span,
+                                   "`let foo;`/`if .. { foo =  ..; }` sequence detected",
```

Comment:
I’m open to suggestions for a good error message and short description (`TODO` is not descriptive enough :stuck_out_tongue_closed_eyes:).